server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -213,43 +213,31 @@ private void validateXfccHeaderAllowed(XfccSourceConfig xfccConfig, Optional<Str
       return Optional.empty();
     }
 
-    // From https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-client-cert:
-    // The XFCC header value is a comma (“,”) separated string. Each substring is an XFCC element,
-    // which holds information added by a single proxy. A proxy can append the current client
-    // certificate information as an XFCC element, to the end of the request’s XFCC header
-    // after a comma.
-    //
-    // Each XFCC element is a semicolon “;” separated string. Each substring is a key-value pair,
-    // grouped together by an equals (“=”) sign. The keys are case-insensitive, the values are
-    // case-sensitive. If “,”, “;” or “=” appear in a value, the value should be double-quoted.
-    // Double-quotes in the value should be replaced by backslash-double-quote (“).
-    String[] xfccElements = xfccHeaderValues.get(0).split(",");
+    // Parse the XFCC header as formatted by Envoy
+    XfccHeader xfccHeader;
+    try {
+      xfccHeader = XfccHeader.parse(xfccHeaderValues.get(0));
+    } catch (XfccHeader.ParseException e) {
+      logger.warn(format("Unable to parse input %s header", XFCC_HEADER_NAME), e);
+      return Optional.empty();
+    }
 
     // Keywhiz currently supports only one certificate set in the XFCC header,
     // since otherwise it is more difficult to distinguish which certificate should have
     // access to secrets
-    if (xfccElements.length == 0) {
+    if (xfccHeader.elements.length == 0) {
       logger.warn("No data provided in {} header", XFCC_HEADER_NAME);
       return Optional.empty();
-    } else if (xfccElements.length > 1) {
+    } else if (xfccHeader.elements.length > 1) {
       logger.warn(
           "Keywhiz only supports one certificate set in the {} header, but {} were provided",
```

Comment:
My concern there is that if the traffic flowed through multiple proxies that each set the XFCC header, then the last set of certs might be for a proxy rather than the intended client.